### ISSUE TITLE
Closing registration for invite-only users

### ISSUE THREAD

**Issue by Braintelligence** [turtl/api#31]
Hi everyone.

Is it possible to close the registration, so people can use your self-hosted API on an invite-only basis with the current config options? (Didn't find anything about this.)

Also, is it possible to call upon any kind of administration to see the usernames/emails of current users and remove them as an admin? (Perhaps via matching of the Admin Email in config and the username?)

**Comment by bashrc** [comment]
Is far as I can tell from models/users.lisp there isn't any limit on new user registrations. Possibly the validator could be modified, although it might need splitting so that editing of existing users still works.

**Comment by hechi** [comment]
this would be really nice because otherwise, my nerdfriends (and i know the will do that ^^ because they can) will spam my turtl and create thousands of users ^^

**Comment by alanmeeson** [comment]
I have a similar use case:  I want to host a turtl server for myself and allow one of my friends to try it out.  Naturally, I don't want the rest of the net to be able to sign up for an account.  

My work around was to put it behind an Apache SSL proxy and add password protection specifically to the /users path[1], since that call is used only for creating a new account I essentially added a password requirement for account creation.

You could adapt this to use a one-time password PAM module for authentication on that one path to allow for a limit on new users.

[1]: I used LocationMatch sections in combination with proxy and virtual host blocks in my apache config.  This let me specify "/users[/]?" and "/users/([0-9a-f]+)" as separate sections with different authentication requirements.

**Comment by marvinwankersteen** [comment]
@alanmeeson 
> [1]: I used LocationMatch sections in combination with proxy and virtual host blocks in my apache config. This let me specify "/users[/]?" and "/users/([0-9a-f]+)" as separate sections with different authentication requirements.

Can you please explain me a little bit more how do you solved this? I just want to use turtl for myself and I dont fully understand your solution.

**Comment by sorcer1122** [comment]
@alanmeeson

I am interested as well. Can you please explain how to achieve this?

**Comment by alanmeeson** [comment]
It's been a while since I did it, and I am rather pressed for time, so this will be a very rough description, I'm afraid I will still have to leave most of this as "an exercise for the reader" so to speak. and I will not be able to offer further help.

Ok, so a little more detail then; 

## The Aim
If you have a look at [users.lisp in the turtl source](https://github.com/turtl/api/blob/master/controllers/users.lisp) you see on line 3 the start of the section of code which handles requests to create new users.  This is invoked by an http call to <your-server-address>/users .  

Lower down you see other blocks that handle things like getting users by id, updating their data, etc..  These are invoked by calling <server-address>/users/([0-9a-f]+).  (Where that last bit is some hexadecimal number, aka combination of numbers and letters from a to f).

We want to block calls to /users, but allow call to /users/([0-9a-f]+)

Also, Turtl makes use of the standard http authentication system for it's logon (i think) so this needs to be passed through too.

## Setting up an Apache reverse proxy
If you are setting up a Turtl server facing the public internet, you probably don't want to be accessing it via http for securities sake,  so it is probably a good idea to set up an SSL proxy.  That is, you connect to the Apache SSL proxy and it forwards you traffic to the otherwise inaccessible Turtl server.

[There are a number of tutorials on the net on how to do this](https://www.digitalocean.com/community/tutorials/how-to-use-apache-http-server-as-reverse-proxy-using-mod_proxy-extension).

## Using the reverse proxy to limit access to /users
Once you have your proxy in place you then are left with the task of preventing people from accessing the /users path, while allowing access to the /users/([0-9a-f]+) path.

[This can be achieved using LocationMatch to password protect the /users/ path](https://tech.enekochan.com/en/2012/09/05/password-protect-a-location-locationmatch-or-virtualhost-in-apache-xampp-and-mac-os-x/).  Beyond that link, figuring out how to set up a password protected location is a task I leave to you to figure out, or google for.

The important thing to remember is that you need to allow access to the /users/([0-9a-f]+) path. or the rest of Turtl will not work either.

I ended up with a section like this somewhere in my apache configuration files.

```#Proxy everything, and don't require a password
<Proxy *>
  AuthType None
  Order allow,deny
  Allow from all
</Proxy>

#Proxy the root of the address, and don't require a password
<Location />
  AuthType None
  Order allow,deny
  Allow from all
  SetEnv proxy-chain-auth On
</Location>

#Anything that wants to access /users must provide a password
<LocationMatch /users[/]?$>
  AuthType Basic
  AuthName "Authentication Required"
  AuthUserFile "/usr/local/apache2/htpasswd/.htpasswd"
  Require valid-user
  SetEnv proxy-chain-auth On
  Order allow,deny
  Allow from all
</LocationMatch>

#Anything that wants to access /users/<useridhere> doesnt need to give a password
<LocationMatch /users/([0-9a-f]+)>
  AuthType None
  Order allow,deny
  Allow from all
  SetEnv proxy-chain-auth On
</LocationMatch>
```

I seem to recall the order in which these sections appeared was important.

I am sorry that is not much to go on, but it should provide you with the key terms to search for and hopefully enough to figure it out and post a better guide here.

**Comment by sorcer1122** [comment]
@alanmeeson Many thanks! You gave me a great start. However, I am having the problem. Let me describe first my nginx config:

location /users {
auth_basic "Blah Blah Restricted Content";
auth_basic_user_file /etc/nginx/.passwd;
}

location /users/36bjy346d2h3xs0081000001 {
allow all;
}

Now, when I am trying to register a new user - I get a password request, this is fine. BUT when I try to login with my main user ID - nothing happens...nothing at all. In the log I see:

POST /auth HTTP/1.1

and this is it!

Would appreciate if anyone could contribute how to fix it?

Thanks

**Comment by alanmeeson** [comment]
The problem is:
> `location /users/36bjy346d2h3xs0081000001`

The above only allows one user rather than all users.  You need to [use a regex here](https://www.scalescale.com/tips/nginx/nginx-location-directive/) as I did:
`/users/([0-9a-f]+)`

**Comment by sorcer1122** [comment]
Yes, yes, this was just for an experiment. Changing it to regex triggers a password request, which should not happen.

**Comment by alanmeeson** [comment]
Hrm... Sadly I can't help with the nginx side of things.

The problems I came across doing this with Apache were:
1. Making the /users location match just that path and not the /users/([0-9a-f]+) subpath.
2. Making the proxy forward on the authentication which turtl uses (that's what the `SetEnv proxy-chain-auth On` was about).

These might be things work looking into.  Good luck.

**Comment by sorcer1122** [comment]
The problem is 2. - nginx does not have something like proxy-chain-auth...so, I guess my problem is that the proxy does not forward authorisation onto Turtl. Digging into this. Thanks for the help.

**Comment by sorcer1122** [comment]
@alanmeeson Thanks once again for your config! I ended up changing nginx to apache and now it works flawlessly! You are  a star!

**Comment by alanmeeson** [comment]
Glad to help @sorcer1122 . While it's fresh in your mind,  would you have time to post a description what steps you had to do to get it to work?  

My post outline the general idea, but as you've just done it you might be able to provide a more detailed "stick this bit in that config file" list of what to do to help the next person to come along with the problem.

**Comment by sorcer1122** [comment]
@alanmeeson Sure, here is what I did:

1. Install apache. There are multiple guides how to do it on Internet, make sure to enable necessary modules (ssl and reverse proxy)

2. I wrote the following config:

Listen xxxx (change xxxx to to apache's external port)
<IfModule mod_ssl.c>
<VirtualHost *:xxxx> (here xxxx is the same as above)
ServerName mydomain.com
SSLEngine On
SSLCertificateFile /path/to/crt
SSLCertificateKeyFile /path/to/key

ProxyPass / http://127.0.0.1:8181/ (this is default turtl's port, change it if you have turtl on a different port)
ProxyPassReserve / http://127.0.0.1:8181/

</VirtualHost>
</IfModule>

3. Write the second config and copy the solution offered by @alanmeeson above.

4. Don't forget to restart apache and modify your firewall's rule to allow your turtl client

5. Enjoy!
